### ISSUE TITLE
[LoRA] Implement Production Training Features

### ISSUE THREAD

**Issue by makr-code** [makr-code/ThemisDB#535]
## 📋 Description

Implement production-ready features for LoRA training including mixed precision, gradient clipping, checkpointing, and distributed training. These features enable efficient, reliable, and scalable training.

**Prerequisites**: Phase 1 & 2 complete (GPU acceleration, llama.cpp integration)  
**Related Issues**: #[Phase 1], #[GPU], #[llama.cpp]

## 🎯 Goals

- [ ] Mixed precision training (FP16/BF16)
- [ ] Gradient accumulation
- [ ] Gradient clipping
- [ ] Learning rate scheduling
- [ ] Checkpointing and resumption
- [ ] Distributed training (multi-GPU)
- [ ] Training monitoring and logging

### 1. Mixed Precision Training
- [ ] FP16 forward/backward passes
- [ ] FP32 master weights and optimizer state
- [ ] Loss scaling to prevent underflow
- [ ] Dynamic loss scaling
- [ ] BF16 support (Brain Float 16)
- [ ] Automatic mixed precision (AMP)

**Benefits**:
- 2x faster training
- 2x less memory
- Maintained accuracy with loss scaling

**Files**:
- `include/llm/lora_framework/mixed_precision.h`
- `src/llm/lora_framework/mixed_precision.cpp`

### 2. Gradient Accumulation
- [ ] Accumulate gradients over N steps
- [ ] Average accumulated gradients
- [ ] Support large effective batch sizes
- [ ] Memory-efficient implementation

**Use Case**: Train with batch_size=32 on GPU with memory for batch_size=4

```cpp
effective_batch_size = micro_batch_size * accumulation_steps
Example: 4 * 8 = 32 effective batch size
```

### 3. Gradient Clipping
- [ ] Clip by global norm
- [ ] Clip by value
- [ ] Configurable threshold
- [ ] Prevent gradient explosion

**Methods**:
```cpp
// Clip by norm
max_norm = 1.0
grad_norm = sqrt(sum(grad_i^2))
if grad_norm > max_norm:
    grad *= max_norm / grad_norm

// Clip by value
grad = clip(grad, -clip_value, clip_value)
```

### 4. Learning Rate Scheduling
- [ ] Warmup (linear)
- [ ] Cosine annealing
- [ ] Cosine with restarts
- [ ] Polynomial decay
- [ ] Step decay
- [ ] Exponential decay
- [ ] Constant with warmup

**Files**:
- `include/llm/lora_framework/lr_scheduler.h`
- `src/llm/lora_framework/lr_scheduler.cpp`

**Example Schedule**:
```
Steps      Learning Rate    Phase
0-100      0 → 1e-4        Warmup
100-1000   1e-4 → 1e-6     Cosine decay
1000+      1e-6            Final
```

### 5. Checkpointing
- [ ] Save training state (model, optimizer, scheduler)
- [ ] Resume from checkpoint
- [ ] Periodic saving (every N steps/epochs)
- [ ] Best model saving (based on validation loss)
- [ ] Checkpoint rotation (keep last K checkpoints)
- [ ] Crash recovery

**Checkpoint Contents**:
```
checkpoint/
├── model_state.bin          # LoRA weights
├── optimizer_state.bin      # Adam moments
├── scheduler_state.bin      # LR schedule state
├── training_state.json      # Step, epoch, metrics
└── metadata.json            # Config, hyperparameters
```

### 6. Distributed Training (Multi-GPU)
- [ ] Data parallelism (each GPU trains on different batch)
- [ ] Gradient synchronization (AllReduce)
- [ ] Multi-node support (optional)
- [ ] Efficient communication (NCCL, Gloo)
- [ ] Load balancing

**Strategies**:
- **Data Parallel**: Most common, good scaling up to 8 GPUs
- **Model Parallel**: For very large models (future)
- **Pipeline Parallel**: For memory-limited setups (future)

### 7. Training Monitoring
- [ ] Real-time metrics (loss, learning rate, throughput)
- [ ] TensorBoard logging
- [ ] Grafana integration (existing infrastructure)
- [ ] Validation metrics
- [ ] Progress bar with ETA
- [ ] Sample generation during training

**Metrics**:
- Training loss (per step)
- Validation loss (per epoch)
- Learning rate (per step)
- Gradient norm (per step)
- Throughput (tokens/sec)
- Memory usage (GPU VRAM)

### 8. Configuration System
- [ ] YAML/JSON configuration files
- [ ] Command-line argument override
- [ ] Configuration validation
- [ ] Preset configurations (quick-start)
- [ ] Documentation of all options

**Example Config**:
```yaml
training:
  max_steps: 10000
  batch_size: 4
  gradient_accumulation_steps: 8
  mixed_precision: "fp16"
  gradient_clipping: 1.0
  
optimizer:
  type: "adamw"
  learning_rate: 1e-4
  weight_decay: 0.01
  
scheduler:
  type: "cosine_with_warmup"
  warmup_steps: 500
  
checkpointing:
  save_every_n_steps: 500
  keep_last_k: 3
  save_best: true
```

### 9. Testing
- [ ] Unit tests for each feature
- [ ] Integration tests (full training pipeline)
- [ ] Multi-GPU tests
- [ ] Checkpoint save/load tests
- [ ] Performance regression tests

## ✅ Acceptance Criteria

- [ ] Mixed precision works and speeds up training
- [ ] Gradient accumulation enables large batch sizes
- [ ] Gradient clipping prevents instability
- [ ] Learning rate scheduling improves convergence
- [ ] Checkpointing allows resuming training
- [ ] Multi-GPU training scales linearly (up to 4-8 GPUs)
- [ ] Monitoring provides useful insights
- [ ] All features configurable via config file
- [ ] Documentation complete

## 🔗 Dependencies

- Phase 1: CPU training
- Phase 2: GPU acceleration, Adam optimizer, llama.cpp
- Existing monitoring infrastructure (Grafana, Prometheus)
- NCCL (NVIDIA) or Gloo (generic) for distributed training

## 📊 Estimated Effort

**Time**: 4-6 weeks  
**Priority**: 🟢 Medium (Phase 3, after core functionality)  
**Complexity**: High (distributed systems, optimization)

## 🧪 Test Strategy

1. **Mixed Precision**: Compare FP16 vs FP32 accuracy and speed
2. **Gradient Accumulation**: Verify batch_size=32 (1x) = batch_size=4 (8x accumulation)
3. **Clipping**: Train with unstable data, verify no NaN gradients
4. **Checkpointing**: Save, kill process, resume, verify identical results
5. **Multi-GPU**: Compare 1 GPU vs 2/4/8 GPUs speedup

### Performance Targets

```
Feature              Speedup      Memory     Accuracy
------------------------------------------------------
Mixed Precision      2x           0.5x       ~same
Gradient Accum       -            1/Nx       same
Multi-GPU (4x)       3.5-4x       same       same
All Combined         ~8x          0.25x      ~same
```

## 📚 References

- Mixed Precision Training: https://arxiv.org/abs/1710.03740
- Distributed Training: https://pytorch.org/tutorials/intermediate/ddp_tutorial.html
- Learning Rate Scheduling: https://arxiv.org/abs/1608.03983
- NCCL Documentation: https://docs.nvidia.com/deeplearning/nccl/

### Mixed Precision Best Practices
- Use FP16 for computation, FP32 for accumulation
- Scale loss by 1024-2048 initially
- Dynamic scaling: reduce on overflow, increase on stability
- Keep master weights in FP32

### Gradient Accumulation Memory Savings
```
Without: batch_size=32 → 8 GB VRAM
With: batch_size=4, accum=8 → 2 GB VRAM (same effective batch)
```

### Distributed Training Scaling
```
GPUs    Speedup (ideal)    Speedup (real)    Efficiency
----------------------------------------------------------
1       1x                 1x                100%
2       2x                 1.9x              95%
4       4x                 3.6x              90%
8       8x                 6.8x              85%
```

### Learning Rate for Distributed
- Scale learning rate with batch size: `lr_new = lr_base * sqrt(batch_size_new / batch_size_base)`
- Example: batch=4, lr=1e-4 → batch=32, lr=2.83e-4

## 🏁 Definition of Done

- [ ] All production features implemented
- [ ] Training stable and reliable
- [ ] Performance targets met
- [ ] Multi-GPU scaling verified
- [ ] Checkpointing robust (crash recovery)
- [ ] Monitoring integrated
- [ ] Documentation complete
- [ ] Ready for production deployment
